specification/under_development/output_schemas/Labeled_And_Async.md

```diff
@@ -1,2 +1,239 @@
+
+### Array Type Records
+
+```llvm
+void @__quantum__rt__array_record_output(i64, i8*)
+```
+
+Inserts a marker in the generated output that indicates the start of an array and how many array elements it has. It produces output records of exactly `"OUTPUT\tARRAY\tn\tlabel"` where `n` is the string representation of the integer value, such as `"OUTPUT\tARRAY\t4\tlabel"`.  The second parameter defines a string label for the result value which is included in the output (`label`).
+
+## Examples
+
+### Basic Single Item Output 
+
+The QIR program would include a single call to the runtime function matching the type of the return value: 
+
+```llvm
+@0 = internal constant [3 x i8] c"0_i\00"
+call void @__quantum__rt__integer_record_output(i64 %5, i8* getelementptr inbounds ([3 x i8], [3 x i8]* @0, i32 0, i32 0))
+ret void
+```
+
+Such that the provider output for `3` shots would be: 
+
+```log
+START
+METADATA\tmetadata1_name_only
+METADATA\tmetadata2_name\tmetadata2_value
+OUTPUT\tINT\t42\t0_i
+END\t0
+START
+METADATA\tmetadata1_name_only
+METADATA\tmetadata2_name\tmetadata2_value
+OUTPUT\tINT\t41\t0_i
+END\t0
+START
+METADATA\tmetadata1_name_only
+METADATA\tmetadata2_name\tmetadata2_value
+OUTPUT\tINT\t42\t0_i 
+END\t0
+```
+
+## Measurement Result Array Output 
+
+For an array of measurement results (or classical values), the QIR program would include an array output recording call, where the first argument indicates the length of the array, followed by the corresponding output recording calls that represent each one of the array items (shown with static result allocation transformations performed):
+```llvm
+@0 = internal constant [5 x i8] c"0_0a\00"
+@1 = internal constant [7 x i8] c"1_0a0r\00"
+@2 = internal constant [5 x i8] c"2_1a\00"
+@3 = internal constant [7 x i8] c"3_1a0r\00"
+@4 = internal constant [7 x i8] c"4_1a1r\00"
+call void @__quantum__rt__array_record_output(i64 1, i8* getelementptr inbounds ([5 x i8], [5 x i8]* @0, i32 0, i32 0))
+call void @__quantum__rt__result_record_output(%Result* %2, i8* getelementptr inbounds ([7 x i8], [7 x i8]* @1, i32 0, i32 0))
+call void @__quantum__rt__array_record_output(i64 2, i8* getelementptr inbounds ([5 x i8], [5 x i8]* @2, i32 0, i32 0))
+call void @__quantum__rt__result_record_output(%Result* nonnull inttoptr (i64 0 to %Result*), i8* getelementptr inbounds ([7 x i8], [7 x i8]* @3, i32 0, i32 0))
+call void @__quantum__rt__result_record_output(%Result* nonnull inttoptr (i64 0 to %Result*), i8* getelementptr inbounds ([7 x i8], [7 x i8]* @4, i32 0, i32 0))
+ret void
+```
+
+This would produce provider output for `3` shots of the form: 
+
+```log
+START
+METADATA\tmetadata1_name_only
+METADATA\tmetadata2_name\tmetadata2_value
+OUTPUT\tARRAY\t1\t0_0a
```

Comment:
Does `t0_0a` have to come before `t2_2a` in the output?

Reply:
@peter-campora I can clarify at the top of the file, for the async and labeled, order is not important, but it puts the responsibility on the party doing the labeling to be able to reconstruct the output structure based on their defined scheme. The usage of `t0_0a` and `t2_2a` are just examples  of a labeling scheme.